eth_abi/registry.py

```diff
@@ -0,0 +1,126 @@
+
+        self._values[predicate] = value
+
+    def find(self, *args, **kwargs):
+        results = tuple(
+            value for predicate, value in self._values.items()
+            if predicate(*args, **kwargs)
+        )
+
+        if len(results) == 0:
+            raise ValueError('No matching values found')
+
+        if len(results) > 1:
+            raise ValueError('Multiple matching values found')
+
+        return results[0]
+
+    def remove_by_reference(self, predicate):
+        # Delete the predicate mapping to the previously stored value
+        try:
+            del self._values[predicate]
+        except KeyError:
+            raise KeyError('Predicate not found')
+
+        # Delete any label which refers to this predicate
+        try:
+            label = self._label_for_predicate(predicate)
+        except ValueError:
+            pass
+        else:
+            del self._labeled_predicates[label]
+
+    def _label_for_predicate(self, predicate):
+        for key, value in self._labeled_predicates.items():
+            if value is predicate:
+                return key
+
+        raise ValueError('Predicate not referred to by any label')
+
+    def remove_by_label(self, label):
+        try:
+            predicate = self._labeled_predicates[label]
+        except KeyError:
+            raise KeyError('Label not found')
+
+        del self._labeled_predicates[label]
+        del self._values[predicate]
+
+    def remove(self, predicate_or_label):
+        if callable(predicate_or_label):
+            self.remove_by_reference(predicate_or_label)
+        elif isinstance(predicate_or_label, str):
+            self.remove_by_label(predicate_or_label)
+        else:
+            raise TypeError('Key to be removed must be a callable or a string')
+
+
+@curry
+def static_predicate(lookup, base, sub, arrlist):
+    return lookup == collapse_type(base, sub, arrlist)
+
+
+class ABIRegistry:
+    def __init__(self):
+        self._encoders = PredicateMapping()
+
+    def register_encoder(self, lookup, encoder, label=None):
+        """
+        Registers an encoder under the given type lookup and optional label.
+        """
+        if callable(lookup):
+            self._encoders.add(lookup, encoder, label)
+            return
+
+        if isinstance(lookup, str):
+            self._encoders.add(static_predicate(lookup), encoder, lookup)
+            return
+
+        raise TypeError('Lookup must be a callable or a value of type `str`')
+
+    def unregister_encoder(self, lookup_or_label):
+        """
+        Unregisters the encoder associated with the given type lookup or label.
+        """
+        if callable(lookup_or_label):
+            try:
+                self._encoders.remove_by_reference(lookup_or_label)
+            except KeyError:
+                raise NoEntryFound('No entry found for given predicate')
+            return
+
+        if isinstance(lookup_or_label, str):
```

Comment:
@pipermerriam Should I maybe be using `eth_utils.is_text` here?

Reply:
I don't think it's needed.  Now that we're in python3 and differentiating between text/bytes doesn't require a check for what python version we're in, I think we can move away from `is_text` and `is_bytes` and potentially even `is_string`.  For lower level libs like this one, reducing the call-stack overhead should yield some basic albeit minor performance gains.